Контроль и ограничения

GoveRaven opened the pull request.

Lexeria commented on project/src/components/genreList.tsx:
```diff
@@ -0,0 +1,37 @@
+  const dispatch = useAppDispatch();
+  const activeGenre = useAppSelector((state) => state.genre);
+
+  films.map((film) => genres.includes(film.genre) || genres.push(film.genre));
```
Подскажи, к чему эта строчка тут? `map` не мутирует исходный массив, а возвращает новый. Получается, что результат никуда не записывается сейчас и сама строчка ради `push` только. А если бы и мутировала, то тоже было бы не очень, так как не очень хорошо мутировать исходные аргументы

Lexeria replied:
Я вижу, какую логику ты пытаешься преследовать, на всякий случай подскажу еще про существование `Set`

GoveRaven replied:
поменял map на forEach

Lexeria commented on project/src/components/genreList.tsx:
```diff
@@ -0,0 +1,37 @@
+            className="catalog__genres-link"
+            onClick={() => {
+              dispatch(changeGenre(genre));
+              dispatch(FilterFilmsByGenre());
```
Название действия с большой буквы стало почему-то, хотя выше с маленькой

Lexeria replied:
Вообще, из-за названия ощущение, что он ожидает `genre` в качестве аргумента. По-хорошему он должен называться как-то в духе `actualizeFilmsListByActiveGenre`, но оно слишком длинное, можно подумать, как сократить, мб просто `actualizeFilmsList`. И будет список последовательных действий читаться круче: поменяй жанр, актуализируй список фильмов

GoveRaven replied:
переименовал

Lexeria commented on project/src/hooks/index.ts:
```diff
@@ -0,0 +1,7 @@
+import { TAppDispatch, TState } from '../types/store';
+import { useSelector } from 'react-redux';
+
+export const useAppDispatch = () => useDispatch<TAppDispatch>();
+
+export const useAppSelector: TypedUseSelectorHook<TState> = useSelector;
```
Это строчки из лайва в принципе, будет круто, если ты поделишься пониманием того, как работает здесь типизация и на что она влияет в коде, который ты пишешь

GoveRaven replied:
на сколько я понимаю, мы задаём джейнерик и тип для хуков, которые предоставляет Redux, и записываем их в переменные. Получаем те же хуки, но они знают о типах данных в нашем хранилище, а поэтому могут подсказывать всё то, что подсказывает TS (куда можно передать значения, какие методы могут быть использованы и т.д.)

Lexeria replied:
В целом да, без тс нам не подсказывают обычно какие экшны можем диспатчить, какой тип вернет стейт и все в этом духе. Благодаря этому ошибиться в мире редакса, где все разбросано по разным частям, будет тяжеловато, а писать код приятнее и быстрее

Lexeria commented on project/src/hooks/index.ts:
```diff
@@ -0,0 +1,7 @@
+import { TypedUseSelectorHook, useDispatch } from 'react-redux';
+import { TAppDispatch, TState } from '../types/store';
+import { useSelector } from 'react-redux';
```
У тебя уже есть импорт из `react-redux` выше, можно убрать эту строчку и дополнить там список

GoveRaven replied:
исправил

Lexeria commented on project/src/index.tsx:
```diff
@@ -2,18 +2,22 @@ import React from 'react';
 import ReactDOM from 'react-dom/client';
 import App from './components/app/app';
 import { films } from './mocks/films';
+import { Provider } from 'react-redux';
+import { store } from './store';
```
Видимо, ребята не завезли линтинг на правила импортов, но просто можешь заметить, что сейчас небольшая каша в порядке импортов и неудобно в случае чего по ним перемещаться. Если не понятно почему — могу попытаться объяснить

GoveRaven replied:
если я правильно понимаю, то сначала нужно писать импорты из библиотек и прочего такого, а потом уже из папок проекта

Lexeria replied:
Не обязательно, это как договариваетесь в команде/проекте, над которым работает. Но единый стиль лучше всегда, чем его отсутствие. Я привык сначала импорт из библиотек, потом из папок уровней выше, а потом в текущем уровне в папке. И все это в алфавитном порядке

Lexeria commented on project/src/index.tsx:
```diff
@@ -2,18 +2,22 @@ import React from 'react';
-      films={films}
-    />
-  </React.StrictMode>
+  <Provider store={store}>
```
Нужно все-таки обернуть только `App`. `React.StrictMode` не будет консьюмером этого провайдера + стоит в `StrictMode` оборачивать все приложение целиком, а не становиться чьим-то ребенком, теряя из вида некоторые компоненты. Зачем? Можно прочитать в [доке](https://react.dev/reference/react/StrictMode)

GoveRaven replied:
исправил

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -199,62 +204,9 @@ export function Main({
+          <GenresList films={films} />
           <div className="catalog__films-list">
-            <FilmList films={films} />
+            <FilmList films={filmsFromStore} />
```
Почему тут `filmsFromStore`, а выше просто `films` из моковых данных? Мб стоит избавиться от этого пропса и перестать прокидывать его из `App`?

GoveRaven replied:
сделал

Lexeria commented on project/src/components/genreList.tsx:
Название файлов лучше писать в едином стиле и `snake-case`

Lexeria replied:
Когда пишут ключевое слово `list` или что-то, что обозначает какое-то множество чего-то, то принято писать во множественном числе — `genres-list`

GoveRaven replied:
переименовал

Lexeria commented on project/src/store/actions.ts:
```diff
@@ -0,0 +1,6 @@
+import { createAction } from '@reduxjs/toolkit';
+
+const changeGenre = createAction<string>('genre/changeGenre');
+const FilterFilmsByGenre = createAction('films/getFilmList');
```
Название переменной поменял, а внутри название (тип) экшна осталось

GoveRaven replied:
сделал

Lexeria commented on project/src/store/actions.ts:
```diff
@@ -0,0 +1,6 @@
+import { createAction } from '@reduxjs/toolkit';
+
+const changeGenre = createAction<string>('genre/changeGenre');
+const FilterFilmsByGenre = createAction('films/getFilmList');
```
В целом, это идейно про одну зону ответственности — фильмы. Так что у обоих можно сделать префикс `films/`

GoveRaven replied:
исправил

Lexeria commented on project/src/store/reducer.ts:
```diff
@@ -0,0 +1,24 @@
+import { films } from '../mocks/films';
+
+const initialState = {
+  genre: 'All genres',
```
Заметь, что у тебя `All genres` дублируется в нескольких местах: на этой строчке, на 16-й, а еще в другом файле (`genreList.tsx`) на 10-й. Так легко очепятаться и сломать логику, советую вытащить в константу и обозвать как-то в духе `DEFAULT_GENRE`

GoveRaven replied:
добавил константу и заменил

Lexeria commented on project/src/types/store.ts:
```diff
@@ -0,0 +1,5 @@
+import { store } from '../store';
+
+export type TState = ReturnType<typeof store.getState>;
+
+export type TAppDispatch = typeof store.dispatch;
```
Тут тоже хочется послушать про понимание типизации, чтобы не упустить этот момент в дальнейшем и самому писать такие типы уметь

GoveRaven replied:
тут у меня такое же объяснение, что и к хукам

Lexeria replied:
Понятен оператор `typeof` и логика вывода типа на основе стора?

GoveRaven committed:
сбросил стили для кнопок жанров и задал ширину для обёртки списка фильмов

GoveRaven committed:
добавил хранилище и обернул проект в него

GoveRaven committed:
добавил компоненту со списком жанров

GoveRaven committed:
добавил компоненту со списком хранилища и поменял аргумент для списка фильмов

GoveRaven committed:
добавил больше фильмов в моки

GoveRaven committed:
переименовал файл

GoveRaven committed:
переименовал переменную действия

GoveRaven committed:
переименовал тип экшена

GoveRaven committed:
переместил импорт хука useSelector выше

GoveRaven committed:
обернут в StrictMode Provider

GoveRaven committed:
изменил префикс у типа экшена

GoveRaven committed:
добавил константу с дефолтным значением жанров и подставил его в код

GoveRaven committed:
убра из компоненты main пропс с моковыми фильмами и заменил их на фильмы из хранилища

GoveRaven committed:
добавил пропс для App с фильмами

GoveRaven committed:
заменил проверку на уникальность жанра на Set

GoveRaven committed:
удалил лишний аргумент

Lexeria commented on project/src/components/app/app.tsx:
```diff
@@ -9,8 +9,11 @@ import { NotFound } from '../../pages/not-found';
 
-type TAppProps = TMainProps;
+type TAppProps = TMainProps & {
+  films: TFilm[]
```
Можно и сейчас убрать, но пока можем оставить. Используется, чтобы прокинуть моки в плеер и на страницу фильма, которые, на самом деле, уже можно прокидывать не через props drilling, а с использованием нашего хранилища в самих компонентах, где они нужны

GoveRaven replied:
заменил на переменную из хранилища

Lexeria commented on project/src/components/genres-list.tsx:
```diff
@@ -1,20 +1,22 @@
 
-  films.map((film) => genres.includes(film.genre) || genres.push(film.genre));
+  films.forEach((film) => genresSet.add(film.genre));
+  const genresArr = Array.from(genresSet);
```
Это не общепринятное сокращение, лучше писать полностью, а можно без этого. Обычно множественное число уже признак того, что это какой-то список

GoveRaven replied:
переименовал

Lexeria commented on project/src/consts/store.ts:
Название файла хорошее, но для явности это больше имеет отношение к жанрам в частности или к фильмам в общем. Можно обозвать `movies.ts` и класть сюда потом всякие константы, связанные с фильмами. Но пока предлагаю просто иметь это в виду, название в целом пока подходит, над масштабируемостью и целью можно задуматься потом

GoveRaven committed:
добавил в хранилище список фильмов, который не изменяется и передал его в качестве пропса компоненту со списком жанров

GoveRaven committed:
поменял map на forEach

GoveRaven committed:
переименовал переменную с массивом жанров

GoveRaven committed:
убрал пропс films
теперь фильмы берутся из хранилища

GoveRaven committed:
убрал проп films
переместил импорт Provider выше

GoveRaven committed:
добавил компоненту кнопки Show more

GoveRaven committed:
удалил разметку кнопки Show more
добавил состояния для количества отображённых фильмов
доабвил функцию для кнопки Show more
Добавил пропс для списка жанров

GoveRaven committed:
добавил пропс функцию, чтобы сбрасывать количество показанных фильмов

Lexeria commented on project/src/components/app/app.tsx:
```diff
@@ -9,39 +9,37 @@ import { NotFound } from '../../pages/not-found';
+import { useAppSelector } from '../../hooks';
 
-type TAppProps = TMainProps;
+type TAppProps = TMainProps
```
Точку с запятой забыл, но скорее отформатировать просто

GoveRaven replied:
добавил

Lexeria commented on project/src/components/showMoreButton.tsx:
```diff
@@ -0,0 +1,17 @@
+type TShowMoreButtonProps = {
+  needHide: boolean;
```
Этот пропс, честно говоря не нужен, мы можем просто в нужном месте не рендерить компонент по какому-то условию, которое ты передаешь в `needHide`:
```
{!needHide && <ShowMoreButton onClickAction={...} />}
```

Только лучше отрицание не делать и упростить условие сразу в нужном ключе, мб назвав переменную с условием  `showMoreButton`, а сам компонент переименовать в `MoreButton` просто

GoveRaven replied:
заменил

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -10,14 +12,15 @@ export type TMainProps = {
   const navigate = useNavigate();
   const filmsFromStore = useAppSelector((state) => state.films);
   const defaultFilmsList = useAppSelector((state) => state.defaultFilmsList);
+  const [lengthOfSshowedFilms, setLengtOfSshowedFilms] = useState(8);
```
Можно попробовать назвать чуть проще в духе `shownCount`. А то сейчас еще `s` дублируется в имени, что тяжело заметить, когда оно длинное

GoveRaven replied:
переименовал

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -10,14 +12,15 @@ export type TMainProps = {
+  const [lengthOfSshowedFilms, setLengtOfSshowedFilms] = useState(8);
+
+  function addFilmsInList() {
+    setLengtOfSshowedFilms(lengthOfSshowedFilms + 8);
```
Заметь, что у тебя число 8 дублируется. Это какое-то параметр, который мы можем захотеть в будущем поменять. Можно вытащить в перменную, но еще лучше утащить в файлик с константами, так как это очень похоже на неизменяемое значение, которое не хочется каждый раз объявлять

GoveRaven replied:
добавил константу `NUMBER_ADDED_MOVIES`

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -202,16 +205,15 @@ export function Main({
-          </div>
+          <GenresList
+            films={defaultFilmsList}
+            onGenreChange={() => (setLengtOfSshowedFilms(8))}
```
Подскажешь, зачем круглые скобки после `=>`?

Lexeria replied:
Есть еще момент про передачу стрелочных функций, но его лучше касаться, когда дойдем до блока с оптимизацией, надеюсь не забуду

GoveRaven replied:
я сам не понял как добавил эти скобки. Убрал

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -202,16 +205,15 @@ export function Main({
+            films={defaultFilmsList}
+            onGenreChange={() => (setLengtOfSshowedFilms(8))}
+          />
+          <FilmList films={filmsFromStore.slice(0, lengthOfSshowedFilms)} />
```
Только сейчас обратил внимание на название. Почему `filmsFromStore`? `defaultFilmsList` тоже со стора ведь, думаю можно убрать лишние пару слов в названии

GoveRaven replied:
переименовал в films

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -202,16 +205,15 @@ export function Main({
+          />
+          <FilmList films={filmsFromStore.slice(0, lengthOfSshowedFilms)} />
+          <ShowMoreButton
+            needHide={lengthOfSshowedFilms >= filmsFromStore.length}
```
Вот это условие можно из пропса вытащить сюда, чтобы получился условный рендеринг. Это, кстати, частая практика, когда мы рисуем что-то условно в духе `booleanStatement && <Component />` — потому что в реакте не рендерятся всякие `null` и булевы значения, что удобно

GoveRaven replied:
уже заменил

Lexeria commented on project/src/components/showMoreButton.tsx:
```diff
@@ -0,0 +1,17 @@
+type TShowMoreButtonProps = {
+  needHide: boolean;
+  onClickAction: VoidFunction;
```
Можно назвать просто `onClick`, ведь это и так подразумевает за собой какое-то действие, которое совершится при клике. В обычно JS тоже есть аналогичное название, а также есть аналогичный служебный пропс при передаче обработчика в реакт-элементы, что ты и делаешь на 12-й строчке. Там без лишних слов, так что и нам нет смысла разглагольствоваться. Исключениями разве что бывают моменты, когда хотим что-то конкретизировать и дать ясности, где она нужна — как со сменой жанров: мы не хотим просто `onChange` передать, ведь так не понятно, а на какой точно `change` надо реагировать. При клике это может быть конкретика — а на что кликают? Тут у нас компонент для кнопки просто, так что вопросов не должно быть, но если элементов в компоненте больше и не понятно, то лучше добавлять ясности: `onIconClick` или `onButtonClick` например

Lexeria commented on project/src/consts/filmList.ts:
Имена должны быть в `snake-case` + можно было перименовать файлик `store.ts` в `films.ts` и туда положить две константны, относящиеся к фильмам

GoveRaven replied:
переименовал и поместил вторую константу туда

Lexeria commented on project/src/consts/filmList.ts:
```diff
@@ -0,0 +1 @@
+export const NUMBER_ADDED_MOVIES = 8;
```
Советую использовать слово `count` вместо `number` и схожих и ставить его в конец. Added — больше про добавленные уже, стоит попробовать подобрать более логичное название, например, `DEFAULT_SHOWN_COUNT`. Оно будет, как логичное значение для init-значения в `useState` + по название подходит и к добавляемым фильмам

GoveRaven replied:
переименовал

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -14,12 +15,13 @@ export type TMainProps = {
   const defaultFilmsList = useAppSelector((state) => state.defaultFilmsList);
-  const [lengthOfSshowedFilms, setLengtOfSshowedFilms] = useState(8);
+  const [shownCount, setShownCount] = useState(NUMBER_ADDED_MOVIES);
+  const showButton = shownCount <= films.length;
```
Тут кнопка одна (но это не точно), но советую давать более ясные названия, чтобы не было путаницы: `showMoreButton`, например

GoveRaven replied:
переименовал

Lexeria commented on project/src/pages/main.tsx:
```diff
@@ -14,12 +15,13 @@ export type TMainProps = {
+  const [shownCount, setShownCount] = useState(NUMBER_ADDED_MOVIES);
+  const showButton = shownCount <= films.length;
 
   function addFilmsInList() {
```
Только сейчас обратил внимание на название. Мб назовем так, чтобы было чуть яснее, что за список подразумевается? Можно так и назвать например `showMoreFilms`

GoveRaven replied:
переименовал

GoveRaven committed:
добавил точку с запятой

GoveRaven committed:
убрал пропс needHide и заменил его на условие

GoveRaven committed:
переименовал lengthOfSshowedFilms в shownCount

GoveRaven committed:
добавил константу для числа добавляемых фильмов

GoveRaven committed:
переименовал переменную filmsFromStore на films

GoveRaven committed:
заменил onClickAction на onClick

GoveRaven committed:
переименовал файл с константами store в films и поместил туда DEFAULT_SHOWN_COUNT

GoveRaven committed:
переименовал NUMBER_ADDED_MOVIES в DEFAULT_SHOWN_COUNT

GoveRaven committed:
переименовал переменную, отвечающую за отображение кнопки Show more

GoveRaven committed:
переименовал функцию для отображения больше фильмов
